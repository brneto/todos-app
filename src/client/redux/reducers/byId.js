import { combineActions, handleActions } from 'redux-actions';
import { flip, prop, identity, merge } from 'ramda';
import { createSelector } from 'reselect';
import { setFetchedTodos, setAddedTodo, setToggledTodo } from '../actions';

// const byId = (state = {}, { payload }) =>
//   payload?.response
//     ? { ...state, ...payload.response.entities.todos }
//     : state;
const initialState = {};
const byId = handleActions(
  {
    [combineActions(setFetchedTodos, setAddedTodo, setToggledTodo)]: {
      next: (state, { payload: { entities } }) => merge(state, entities.todos),
    }
  },
  initialState
);

export default byId;

// SELECTORS
// flip(prop) :: {s: a} -> s -> a | Undefined
export const createGetTodo = createSelector([flip(prop)], identity);
