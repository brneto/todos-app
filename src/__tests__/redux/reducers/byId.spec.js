import deepFreeze from 'deep-freeze';
import byId from '../../../client/redux/reducers/byId';
import {
  setFetchTodos,
  setAddTodo,
  setToggleTodo
} from '../../../client/redux/actions';

describe('reducers/byId', () => {
  const filter = 'all';
  const initialState = deepFreeze(byId(undefined, { type: 'INIT' }));
  const baseData = {
    entities: {
      todos: {
        1: {
          id: 1,
          text: 'hey',
          completed: true
        },
        2: {
          id: 2,
          text: 'ho',
          completed: true
        },
        3: {
          id: 3,
          text: 'let\'s go',
          completed: false
        }
      }
    }
  };
  const baseState = deepFreeze(
    byId(initialState, setFetchTodos(baseData, filter))
  );

  it('should handle unknown actions', () => {
    expect(byId(initialState, { type: 'FAKE' })).toBe(initialState);
  });

  it('should add fetched todos to empty map', () => {
    expect(
      byId(initialState, setFetchTodos(baseData, filter))
    ).toMatchSnapshot();
  });

  it('should add fetched todos to non-empty map', () => {
    const testData = {
      entities: {
        todos: {
          4: {
            id: 4,
            text: 'wow',
            completed: false
          },
          5: {
            id: 5,
            text: 'cool',
            completed: true
          }
        }
      }
    };

    expect(byId(baseState, setFetchTodos(testData, filter))).toMatchSnapshot();
  });

  it('should add todo to empty map', () => {
    const testData = {
      entities: {
        todos: {
          4: {
            id: 4,
            text: 'wow',
            completed: false
          }
        }
      }
    };

    expect(byId(initialState, setAddTodo(testData))).toMatchSnapshot();
  });

  it('should add todo to non-empty map', () => {
    const testData = {
      entities: {
        todos: {
          4: {
            id: 4,
            text: 'wow',
            completed: false
          }
        }
      }
    };

    expect(byId(baseState, setAddTodo(testData))).toMatchSnapshot();
  });

  it('should toggle todo to non-empty map', () => {
    const testData = {
      entities: {
        todos: {
          1: {
            id: 1,
            text: 'hey',
            completed: false
          }
        }
      }
    };

    expect(byId(baseState, setToggleTodo(testData))).toMatchSnapshot();
  });
});
