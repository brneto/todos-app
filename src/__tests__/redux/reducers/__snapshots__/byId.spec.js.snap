// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/byId should add fetched todos to empty map 1`] = `
Object {
  "1": Object {
    "completed": true,
    "id": 1,
    "text": "hey",
  },
  "2": Object {
    "completed": true,
    "id": 2,
    "text": "ho",
  },
  "3": Object {
    "completed": false,
    "id": 3,
    "text": "let's go",
  },
}
`;

exports[`reducers/byId should add fetched todos to non-empty map 1`] = `
Object {
  "1": Object {
    "completed": true,
    "id": 1,
    "text": "hey",
  },
  "2": Object {
    "completed": true,
    "id": 2,
    "text": "ho",
  },
  "3": Object {
    "completed": false,
    "id": 3,
    "text": "let's go",
  },
  "4": Object {
    "completed": false,
    "id": 4,
    "text": "wow",
  },
  "5": Object {
    "completed": true,
    "id": 5,
    "text": "cool",
  },
}
`;

exports[`reducers/byId should add todo to empty map 1`] = `
Object {
  "4": Object {
    "completed": false,
    "id": 4,
    "text": "wow",
  },
}
`;

exports[`reducers/byId should add todo to non-empty map 1`] = `
Object {
  "1": Object {
    "completed": true,
    "id": 1,
    "text": "hey",
  },
  "2": Object {
    "completed": true,
    "id": 2,
    "text": "ho",
  },
  "3": Object {
    "completed": false,
    "id": 3,
    "text": "let's go",
  },
  "4": Object {
    "completed": false,
    "id": 4,
    "text": "wow",
  },
}
`;

exports[`reducers/byId should toggle todo to non-empty map 1`] = `
Object {
  "1": Object {
    "completed": false,
    "id": 1,
    "text": "hey",
  },
  "2": Object {
    "completed": true,
    "id": 2,
    "text": "ho",
  },
  "3": Object {
    "completed": false,
    "id": 3,
    "text": "let's go",
  },
}
`;
